[JSC] Do not destroy VM after clearing AtomStringTable for current thread

Constellation opened the pull request:
#### fa7db09c3bb2ba6f4263dce4b0d55aad85e02260
<pre>
[JSC] Do not destroy VM after clearing AtomStringTable for current thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=268415">https://bugs.webkit.org/show_bug.cgi?id=268415</a>
<a href="https://rdar.apple.com/86151259">rdar://86151259</a>

Reviewed by Michael Saboff.

This patch changes the ordering of VM destruction and thread&apos;s AtomStringTable clearing.
This happens when JSVirtualMachine gets destroyed during execution of microtasks. While this
should not happen (because it is destroying VM while running code associated to this VM), we
can alleviate this case by destroying VM under the right AtomStringTable.

* Source/JavaScriptCore/runtime/JSLock.cpp:
(JSC::JSLock::willReleaseLock):

Canonical link: <a href="https://commits.webkit.org/273803@main">https://commits.webkit.org/273803@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d8c975965b2b6e9199a099d7eec6117a7f733efe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36693 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15631 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38924 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39341 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/32893 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18109 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12744 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31470 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37255 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13173 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32441 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11525 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11542 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32705 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40588 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/30794 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33237 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33054 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37464 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/36456 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11814 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9638 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35584 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13484 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/43253 "Built successfully") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/12219 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/8946 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4755 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12701 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

msaboff reviewed (approved):
r=me if EWS is clear

Constellation committed:
[JSC] Do not destroy VM after clearing AtomStringTable for current thread
https://bugs.webkit.org/show_bug.cgi?id=268415
rdar://86151259

Reviewed by Michael Saboff.

This patch changes the ordering of VM destruction and thread's AtomStringTable clearing.
This happens when JSVirtualMachine gets destroyed during execution of microtasks. While this
should not happen (because it is destroying VM while running code associated to this VM), we
can alleviate this case by destroying VM under the right AtomStringTable.

* Source/JavaScriptCore/runtime/JSLock.cpp:
(JSC::JSLock::willReleaseLock):

Canonical link: https://commits.webkit.org/273803@main
